File: packages/push/ios/PushSerializers.swift
/*
 * Copyright (c) 2026 Ping Identity Corporation. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */
import PingPush
import RNPingCore

// MARK: - Credential serialization

/// Converts a native `PushCredential` to a bridge-safe dictionary.
///
/// The `sharedSecret` field is intentionally omitted to avoid leaking
/// sensitive key material across the RN bridge.
func serializeCredential(_ cred: PushCredential) -> [String: Any] {
  var dict: [String: Any] = [:]
  dict["id"] = cred.id
  dict["userId"] = cred.userId as Any? ?? NSNull()
  dict["resourceId"] = cred.resourceId as Any? ?? NSNull()
  dict["issuer"] = cred.issuer
  dict["displayIssuer"] = cred.displayIssuer as Any? ?? NSNull()
  dict["accountName"] = cred.accountName
  dict["displayAccountName"] = cred.displayAccountName as Any? ?? NSNull()
  dict["serverEndpoint"] = cred.serverEndpoint as Any? ?? NSNull()
  dict["createdAt"] = cred.createdAt.timeIntervalSince1970 * 1000
  dict["imageURL"] = cred.imageURL as Any? ?? NSNull()
  dict["backgroundColor"] = cred.backgroundColor as Any? ?? NSNull()
  dict["policies"] = cred.policies as Any? ?? NSNull()
  dict["lockingPolicy"] = cred.lockingPolicy as Any? ?? NSNull()
  dict["isLocked"] = cred.isLocked
  dict["platform"] = serializePlatform(cred.platform)
  // sharedSecret NEVER included — bridge security contract
  return dict
}

/// Reconstructs a `PushCredential` from a bridge dictionary.
///
/// - Throws: `NSError` (via `ReadableMapUtils`) when `id`, `issuer`, or `accountName` is absent.
func deserializeCredential(_ dict: NSDictionary) throws -> PushCredential {
  let id = try ReadableMapUtils.requireString(dict, key: "id")
  let issuer = try ReadableMapUtils.requireString(dict, key: "issuer")
  let accountName = try ReadableMapUtils.requireString(dict, key: "accountName")

  // TODO-PUSH-PARITY: sharedSecret is non-optional on Android PushCredential but not required
  // here on iOS — Android SDK should make it optional so saveCredential round-trips work without
  // a placeholder empty string.
  return PushCredential(
    id: id,
    userId: dict["userId"] as? String,
    resourceId: dict["resourceId"] as? String,
    issuer: issuer,
    displayIssuer: dict["displayIssuer"] as? String,
    accountName: accountName,
    displayAccountName: dict["displayAccountName"] as? String,
    serverEndpoint: dict["serverEndpoint"] as? String ?? "",
    sharedSecret: "",
    createdAt: Date(timeIntervalSince1970: ((dict["createdAt"] as? NSNumber)?.doubleValue ?? 0) / 1000),
    imageURL: dict["imageURL"] as? String,
    backgroundColor: dict["backgroundColor"] as? String,
    policies: dict["policies"] as? String,
    lockingPolicy: dict["lockingPolicy"] as? String,
    isLocked: dict["isLocked"] as? Bool ?? false,
    platform: parsePlatform(dict["platform"] as? String)
  )
}

// MARK: - Notification serialization

/// Converts a native `PushNotification` to a bridge-safe dictionary.
func serializeNotification(_ n: PushNotification) -> [String: Any] {
  var dict: [String: Any] = [:]
  dict["id"] = n.id
  dict["credentialId"] = n.credentialId
  dict["ttl"] = n.ttl
  dict["messageId"] = n.messageId
  dict["messageText"] = n.messageText as Any? ?? NSNull()
  dict["customPayload"] = n.customPayload as Any? ?? NSNull()
  dict["challenge"] = n.challenge as Any? ?? NSNull()
  dict["numbersChallenge"] = n.numbersChallenge as Any? ?? NSNull()
  dict["loadBalancer"] = n.loadBalancer as Any? ?? NSNull()
  dict["contextInfo"] = n.contextInfo as Any? ?? NSNull()
  dict["pushType"] = n.type
  dict["createdAt"] = n.createdAt.timeIntervalSince1970 * 1000
  dict["sentAt"] = n.sentAt.map { $0.timeIntervalSince1970 * 1000 } as Any? ?? NSNull()
  dict["respondedAt"] = n.respondedAt.map { $0.timeIntervalSince1970 * 1000 } as Any? ?? NSNull()
  dict["approved"] = n.approved
  dict["pending"] = n.pending
  return dict
}

// MARK: - Platform

/// Maps an optional JS platform string (`"PING_AM"`, `"PING_ONE"`) to `PushPlatform`.
///
/// Unrecognized or absent values fall back to `.pingAM`.
func parsePlatform(_ value: String?) -> PushPlatform {
  // Only one platform exists in the native SDK currently; default to it.
  return .pingAM
}

/// Maps a native `PushPlatform` to the JS contract string (`"PING_AM"`, `"PING_ONE"`).
func serializePlatform(_ platform: PushPlatform) -> String {
  switch platform {
  case .pingAM:
    return "PING_AM"
  }
}

File: packages/push/android/src/main/java/com/pingidentity/rnpush/PushSerializers.kt
/*
 * Copyright (c) 2026 Ping Identity Corporation. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */
package com.pingidentity.rnpush

import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.ReadableMap
import com.facebook.react.bridge.WritableMap
import com.pingidentity.mfa.push.PushCredential
import com.pingidentity.mfa.push.PushNotification
import com.pingidentity.rncore.utils.requireString
import java.util.Date

// ---------------------------------------------------------------------------
// PushCredential serialization
// ---------------------------------------------------------------------------

/**
 * Converts a native [PushCredential] to a React Native [WritableMap] for bridge transport.
 *
 * The bridge representation uses keys that match the TypeScript `PushCredential` type
 * defined in `@ping-identity/rn-push`. Android SDK field names are mapped to their
 * corresponding bridge keys where they differ (e.g. `timeAdded` → `"createdAt"`).
 *
 * The `sharedSecret` field is deliberately excluded — it is a security-sensitive value
 * kept native-side only and must never cross the bridge.
 *
 * @param credential The native push credential to serialize.
 * @return A [WritableMap] with all bridgeable credential fields using bridge-standard keys.
 */
internal fun serializeCredential(credential: PushCredential): WritableMap {
    val map = Arguments.createMap()
    map.putString("id", credential.id)
    map.putString("issuer", credential.issuer)
    map.putString("accountName", credential.accountName)
    // TODO-PUSH-PARITY: displayIssuer and displayAccountName are non-nullable on Android but
    // nullable on iOS — Android SDK should align to nullable to match the bridge contract.
    map.putString("displayIssuer", credential.displayIssuer)
    map.putString("displayAccountName", credential.displayAccountName)
    // createdAt is a Date on the published AAR
    map.putDouble("createdAt", credential.createdAt.time.toDouble())
    // policies is a nullable JSON string
    if (credential.policies != null) {
        map.putString("policies", credential.policies)
    } else {
        map.putNull("policies")
    }
    // lockingPolicy is a nullable JSON string (null when not set on the SDK)
    if (credential.lockingPolicy != null) {
        map.putString("lockingPolicy", credential.lockingPolicy)
    } else {
        map.putNull("lockingPolicy")
    }
    map.putBoolean("isLocked", credential.isLocked)
    // imageURL is a nullable display hint
    if (credential.imageURL != null) {
        map.putString("imageURL", credential.imageURL)
    } else {
        map.putNull("imageURL")
    }
    // backgroundColor is a nullable display hint
    if (credential.backgroundColor != null) {
        map.putString("backgroundColor", credential.backgroundColor)
    } else {
        map.putNull("backgroundColor")
    }
    // TODO-PUSH-PARITY: platform is a raw String on Android but a typed PushPlatform enum on iOS
    // — Android SDK should expose a typed enum to match.
    map.putString("platform", credential.platform)
    // TODO-PUSH-PARITY: userId, resourceId, serverEndpoint are present on the iOS PushCredential
    // but missing from the Android PushCredential SDK class — emit null as a placeholder.
    map.putNull("userId")
    map.putNull("resourceId")
    map.putNull("serverEndpoint")
    // sharedSecret is intentionally NOT included — never crosses the bridge
    return map
}

/**
 * Reconstructs a native [PushCredential] from a React Native [ReadableMap].
 *
 * The `sharedSecret` field is not present in the bridge map since it is never
 * serialized to JS. A placeholder empty string is used to satisfy the non-optional
 * constructor requirement of the native Android SDK.
 *
 * @param map The bridge map containing credential fields.
 * @return A native [PushCredential] reconstructed from the bridge fields.
 * @throws IllegalArgumentException when required string fields are missing or blank.
 */
internal fun deserializeCredential(map: ReadableMap): PushCredential {
    val id = requireString(map, "id")
    val issuer = requireString(map, "issuer")
    val accountName = requireString(map, "accountName")
    val displayIssuer = optionalStringField(map, "displayIssuer")
    val displayAccountName = optionalStringField(map, "displayAccountName")
    // "createdAt" (bridge key) → timeAdded (Android SDK field)
    val timeAdded = if (map.hasKey("createdAt") && !map.isNull("createdAt")) {
        map.getDouble("createdAt").toLong()
    } else {
        0L
    }
    val policies = optionalStringField(map, "policies")
    val lockingPolicy = optionalStringField(map, "lockingPolicy")
    val isLocked = if (map.hasKey("isLocked")) map.getBoolean("isLocked") else false
    val imageURL = optionalStringField(map, "imageURL")
    val backgroundColor = optionalStringField(map, "backgroundColor")
    val platformStr = if (map.hasKey("platform") && !map.isNull("platform")) {
        map.getString("platform") ?: "PING_AM"
    } else {
        "PING_AM"
    }

    // TODO-PUSH-PARITY: sharedSecret is non-optional on Android but optional on iOS — Android SDK
    // should make it optional so saveCredential round-trips work without a placeholder empty string.
    return PushCredential(
        id = id,
        issuer = issuer,
        accountName = accountName,
        displayIssuer = displayIssuer ?: issuer,
        displayAccountName = displayAccountName ?: accountName,
        createdAt = Date(timeAdded),
        sharedSecret = "",
        serverEndpoint = "",
        policies = policies,
        lockingPolicy = lockingPolicy,
        isLocked = isLocked,
        imageURL = imageURL,
        backgroundColor = backgroundColor,
        platform = platformStr,
    )
}

// ---------------------------------------------------------------------------
// PushNotification serialization
// ---------------------------------------------------------------------------

/**
 * Converts a native [PushNotification] to a React Native [WritableMap] for bridge transport.
 *
 * Notification objects travel native → JS only; there is no deserialization path.
 *
 * The bridge representation uses keys that match the TypeScript `PushNotification` type.
 * Android SDK field names are mapped to their bridge keys where they differ
 * (e.g. `type` → `"pushType"`, `message` → `"messageText"`, `timeAdded` → `"createdAt"`).
 *
 * @param notification The native push notification to serialize.
 * @return A [WritableMap] with all bridgeable notification fields using bridge-standard keys.
 */
internal fun serializeNotification(notification: PushNotification): WritableMap {
    val map = Arguments.createMap()
    map.putString("id", notification.id)
    map.putString("credentialId", notification.credentialId)
    map.putDouble("ttl", notification.ttl.toDouble())
    // createdAt is a Date on the published AAR
    map.putDouble("createdAt", notification.createdAt.time.toDouble())
    notification.sentAt?.let { map.putDouble("sentAt", it.time.toDouble()) } ?: map.putNull("sentAt")
    notification.respondedAt?.let { map.putDouble("respondedAt", it.time.toDouble()) } ?: map.putNull("respondedAt")
    // messageText (published AAR field name)
    if (notification.messageText != null) {
        map.putString("messageText", notification.messageText)
    } else {
        map.putNull("messageText")
    }
    if (notification.customPayload != null) {
        map.putString("customPayload", notification.customPayload)
    } else {
        map.putNull("customPayload")
    }
    // TODO-PUSH-PARITY: challenge field is present on iOS PushNotification but missing from
    // Android PushNotification SDK class — emit null as a placeholder.
    map.putNull("challenge")
    if (notification.numbersChallenge != null) {
        map.putString("numbersChallenge", notification.numbersChallenge)
    } else {
        map.putNull("numbersChallenge")
    }
    map.putString("loadBalancer", notification.loadBalancer)
    map.putString("contextInfo", notification.contextInfo)
    map.putString("messageId", notification.messageId)
    map.putString("pushType", notification.pushType.name.lowercase())
    map.putBoolean("approved", notification.approved)
    map.putBoolean("pending", notification.pending)
    return map
}

// ---------------------------------------------------------------------------
// Private helpers
// ---------------------------------------------------------------------------

/**
 * Reads an optional nullable string value from a React Native [ReadableMap].
 *
 * @param map The map to read from.
 * @param key The key to look up.
 * @return The string value, or `null` when the key is absent or null.
 */
private fun optionalStringField(map: ReadableMap, key: String): String? {
    return if (map.hasKey(key) && !map.isNull(key)) map.getString(key) else null
}
